Source/WebCore/rendering/svg/RenderSVGEllipse.cpp

```diff
@@ -62,6 +62,11 @@ void RenderSVGEllipse::updateShapeFromElement()
     if (m_radii.isEmpty())
         return;
 
+    if (m_radii.width() == m_radii.height())
+        m_shapeType = ShapeType::Circle;
+    else
+        m_shapeType = ShapeType::Ellipse;
```

Comment:
This `shapeType` is really weird. I looked at the complete patch in https://github.com/WebKit/WebKit/pull/18444 and I did not find any difference in treating the `Circle` from the `Ellipse`. They are always treated the same way. So why we have to differentiate between them here? And why do we even bother by having `m_shapeType`?

Can't we have a virtual method like `RenderSVGShape::canUseStrokeApproximation()` `RenderSVGShape::isSimpleStroke()` or `RenderSVGShape::hasSimpleStroke()` with the default value false, and make `RenderSVGEllipse` and `RenderSVGRect` return true for it?

Reply:
This is because the shapeType for Ellipse and Circle is the critical part of the hit testing optimization. Right now we are always asking to path when there is a path. But if the stroke is smooth and it is a circle, we do not need to use path-based hit-testing (we can see whether the hit-testing-point from the center of the circle is within the stroke-width of the circle).

Reply:
My question is why do not we have `ShapeType::Ellipse` and get rid of `ShapeType::Circle'? So this code can be 
```
m_shapeType = ShapeType::Ellipse;
```
Alternatively we define this function 
```
class RenderSVGShape : public RenderSVGModelObject {
     virtual bool canUseStrokeApproximation() { return false; }
};
```
And override it for RenderSVGEllipse:
```
class RenderSVGEllipse final : public RenderSVGShape  {
     bool canUseStrokeApproximation() override { return true; }
};
```

Reply:
For `canUseStrokeApproximation`, this is not necessary to make it a virtual function because `shapeDependentStrokeContains` is already a virtual function. We can define this function in the derived class without virtual since we will have the fast path in each derived class.
We will define `bool RenderSVGEllipse::canUseStrokeApproximation() const` and use it in `RenderSVGEllipse::shapeDependentStrokeContains`.

And when using `ShapeType` etc. for the fast path, it is not approximation since this is not a repaint. So probably we will define it like, `canUseStrokeContainsFastPath`

And when defining this function, it is useful to have this `ShapeType::Circle` because this tells quickly that we can use the optimization or not.

So, then inside `RenderSVGEllipse::shapeDependentStrokeContains`

```
if (!canUseStrokeContainsFastPath()) {
    ensurePath();
    return RenderSVGShape::shapeDependentStrokeContains(...);
}

... doing the fast path ...
```

And then, implementation of `canUseStrokeContainsFastPath` will be,

```
bool RenderSVGEllipse::canUseStrokeContainsFastPath() const  // Not virtual function.
{
    if (hasNonScalingStroke())
        return false;
    return m_shapeType == ShapeType::Circle && style().svgStyle().strokeDashArray().isEmpty();
}
```

And we will use `ShapeType::Circle`. This circle shape-type makes it clear the intent of this fast path. When it is circle and it is not a dash stroke, then we can compute the fast path.

Then, why we cannot use the fast path with Ellipse too and just always having Ellipse instead of adding Circle is that the current fast path has an issue, and this one should work only with Circle case. This is https://bugs.chromium.org/p/chromium/issues/detail?id=393143, and the hittest is included in WPT, and we are right now failing because we are using fast path for Ellipse case too. https://wpt.fyi/results/svg/interact/scripted/ellipse-hittest.html?label=experimental&label=master&aligned
The reason why the current fast path does not work is that we are assuming that half-stroke-width is enough to compute hit-test with radius from the center-point. But this only works with Circle and this does not work with Ellipse.